doc/storage/api/api.rst

```diff
@@ -649,24 +691,193 @@ These definitions must be defined in the header file :file:`psa/protected_storag
+
+   Except in the case of ``PSA_ERROR_STORAGE_FAILURE``, in which case no guarantees can be made, the operation shall either succeed or leave storage unchanged. 
+
+
+.. function:: psa_ps_iterator_start
+
+   .. summary:: 
+       Initialises an iterator that can be used to return a list of uids in the Protected Storage .
+
+   .. param::  psa_storage_iterator_t* context
+       A pointer to a context for this iterator. The pointer may be NULL. This is set to a new value on success and is undefined on error. The content of the iterator is implementation defined.    
+       
+   .. param:: psa_storage_uid_t filter
+       A value used to filter the results included in this iteration. 
+       
+   .. param:: int_t filter_length
+       A length of the filter to use, this must be a value `0 < filter_lemngth < 63`. 
+
+   .. param:: psa_storage_uid_t *result
+        A pointer to the location in which to store ``uid``. On success the contents of this location will be updated with the first matching ``uid``. On error, the contents are undefined. 
+        
+   .. return:: psa_status_t
+       A status indicating the success or failure of the operation.
+
+   .. retval:: PSA_SUCCESS
+       The operation completed successfully.
+       
+   .. retval:: PSA_ERROR_DOES_NOT_EXIST
+       No ``uid`` matches this iteration.
+
+   .. retval:: PSA_ERROR_STORAGE_FAILURE
+       The operation failed because the physical storage has failed (Fatal error).
+
+   The iterator returns those values where the ``filter_length`` bits of the `uid` matches the left most bits in ``filter``. 
+
+   The iterator will only returns those ``uid`` that were created by the caller. It MUST not return any ``uid`` created by a different user. 
+
+   An iterator is not required to return uids in any specific order, but MUST return them in a consistent order each time it is called. For example, if an implementation returns entries in numerical order, it should not arbitrarily change to returning them in creation order. However, the caller should not make assumptions as to the order in which entries are returned, except that each uid will be returned only once in each iteration.
+
+   Changes to storage by other users MUST NOT affect any open iterations. 
+
+   A caller may initialize multiple iteration contexts at the same time. Each iteration shall be independent. Calling ``psa_ps_iterator_next()`` on one iterator MUST not effect any other open iteration. 
+   
+   An iterator MUST return all data objects whose `uid` matches the filter that are extant when the filter was created, unless these are deleted or renamed before the iteration would return them, or the caller stops before all matching objects have been returned. 
+   
+   A caller may delete a `uid` with ``psa_ps_remove(uid)`` without invalidating the iteration context. the iterator MUST never return a `uid` that has been deleted. However, if the caller is multi-threaded it ia possible another thread may delete a `uid`. 
```

Comment:
* For inline code that includes API identifiers, using the `:code:` role enables the identifiers to be hyperlinked.
* For inline code that is _not_ an API element, use double back-quotes.
* For inline code that is _only_ an API element, use single back-quotes.

If in doubt, use the `:code:` role which will hyperlink if possible, otherwise render as code.

```suggestion
   A caller may delete a ``uid`` with :code:`psa_ps_remove(uid)` without invalidating the iteration context. the iterator MUST never return a ``uid`` that has been deleted. However, if the caller is multi-threaded it ia possible another thread may delete a ``uid``. 
```